src/sallyport/mod.rs

```diff
@@ -195,32 +197,95 @@ impl<'a> Cursor<'a> {
     pub unsafe fn alloc<T>(
         self,
         count: usize,
-    ) -> core::result::Result<(Cursor<'a>, &'a mut [T]), ()> {
+    ) -> core::result::Result<(Cursor<'a>, &'a mut [MaybeUninit<T>]), ()> {
         let mid = {
-            let (padding, data, _) = self.0.align_to_mut::<T>();
+            let (padding, data, _) = self.0.align_to_mut::<MaybeUninit<T>>();
 
             if data.len() < count {
                 return Err(());
             }
 
-            padding.len() + size_of::<T>() * count
+            padding.len() + size_of::<MaybeUninit<T>>() * count
         };
 
         let (data, next) = self.0.split_at_mut(mid);
 
-        Ok((Cursor(next), data.align_to_mut::<T>().1))
+        Ok((Cursor(next), data.align_to_mut::<MaybeUninit<T>>().1))
     }
 
-    /// Copies data from a value into a slice using self.alloc().
+    /// Copies data from a slice into a slice using self.alloc().
     #[allow(dead_code)]
-    pub fn copy_slice<T: Copy>(
+    pub fn copy_from_slice<T: 'a + Copy>(
         self,
-        value: &[T],
+        src: &[T],
     ) -> core::result::Result<(Cursor<'a>, &'a mut [T]), ()> {
-        let (c, slice) = unsafe { self.alloc(value.len())? };
-        slice.copy_from_slice(value);
+        let (c, slice) = unsafe { self.alloc::<T>(src.len())? };
+        let slice = unsafe { &mut *(slice as *mut [MaybeUninit<T>] as *mut [T]) };
+
+        // FIXME: should be volatile
+        slice.copy_from_slice(src);
+
         Ok((c, slice))
     }
+
+    /// Copies data into a slice from a slice using self.alloc().
+    #[allow(dead_code)]
+    pub fn copy_into_slice<T: 'a + Copy>(
+        self,
+        dst: &mut [T],
+    ) -> core::result::Result<Cursor<'a>, ()> {
+        let (c, slice) = unsafe { self.alloc::<T>(dst.len())? };
+        let slice = unsafe { &mut *(slice as *mut [MaybeUninit<T>] as *mut [T]) };
+
+        // FIXME: should be volatile
+        dst.copy_from_slice(slice);
+
+        Ok(c)
+    }
+
+    /// Reads data from the cursor using self.alloc() and advancing the cursor.
+    #[allow(dead_code)]
+    pub fn read_volatile<T: 'a + Copy>(self) -> core::result::Result<(Cursor<'a>, T), ()> {
+        let (c, slice) = unsafe { self.alloc::<T>(1)? };
+
+        let src = slice[0].as_ptr();
+        let mut dst = MaybeUninit::<T>::uninit();
```

Comment:
`uninit()` is basically `new()` for `MaybeUninit<T>`. You cannot create `MaybeUninit<T>` yourself. You have to use either `uninit()` or `zeroed()` or something else. `MaybeUninit::<T>::uninit()` is basically `malloc(sizeof(T))`, where the contents of the allocated memory is undefined, which does not matter here, because we fill it with ` dst.as_mut_ptr().write()` right afterwards.